File: search/array-based-search/hash-search/python/test_hash_search.py
import unittest
from hash_search import HashTable


class TestHashTable(unittest.TestCase):
    
    def setUp(self):
        self.ht = HashTable(initial_size=4)
    
    def test_insert_and_search(self):
        self.ht.insert("isbn1", {"title": "Book 1"})
        result = self.ht.search("isbn1")
        self.assertIsNotNone(result)
        self.assertEqual(result["title"], "Book 1")
    
    def test_update_existing_key(self):
        self.ht.insert("isbn1", {"title": "Old Title"})
        self.ht.insert("isbn1", {"title": "New Title"})  # Update
        result = self.ht.search("isbn1")
        self.assertEqual(result["title"], "New Title")
    
    def test_delete(self):
        self.ht.insert("isbn1", {"title": "To Delete"})
        self.assertTrue("isbn1" in self.ht)
        
        self.ht.delete("isbn1")
        self.assertIsNone(self.ht.search("isbn1"))
        self.assertFalse("isbn1" in self.ht)
    
    def test_search_missing_key(self):
        result = self.ht.search("missing")
        self.assertIsNone(result)
    
    def test_load_factor_triggers_resize(self):
        # Fill up to trigger resize (size=4 → resize at ~3 items)
        books = [f"isbn{i}" for i in range(5)]
        for isbn in books:
            self.ht.insert(isbn, {"title": f"Book {isbn}"})
        
        # After inserts, size should have doubled
        self.assertGreater(self.ht.size, 4)
        self.assertGreaterEqual(self.ht.count, 5)
    
    def test_collision_handling_separate_chaining(self):
        # Force collision by using keys with same hash (rare, but test structure)
        self.ht.insert("a", {"title": "A"})
        self.ht.insert("b", {"title": "B"})
        
        # Both may go to same bucket — check that both are retrievable
        self.assertEqual(self.ht.search("a")["title"], "A")
        self.assertEqual(self.ht.search("b")["title"], "B")
    
    def test_len_and_contains(self):
        self.ht.insert("k1", {"x": 1})
        self.ht.insert("k2", {"x": 2})
        
        self.assertEqual(len(self.ht), 2)
        self.assertIn("k1", self.ht)
        self.assertIn("k2", self.ht)
        self.assertNotIn("k3", self.ht)
    
    def test_empty_table(self):
        self.assertEqual(len(self.ht), 0)
        self.assertEqual(self.ht.count, 0)
        self.assertEqual(self.ht.load_factor, 0.0)
        self.assertIsNone(self.ht.search("any"))


if __name__ == "__main__":
    unittest.main()

File: search/array-based-search/hash-search/javascript/test_hash_search.js
const { strictEqual, ok, deepEqual } = require('assert');

// Import the class
const { HashTable } = require('./hash_search');

console.log("🧪 Running Hash Search Tests...\n");

// Test: Insert and search
(() => {
    const ht = new HashTable(4);
    ht.insert("isbn1", { title: "Book 1" });
    const result = ht.search("isbn1");
    ok(result, "Should find inserted book");
    strictEqual(result.title, "Book 1", "Title should match");
})();

// Test: Update existing key
(() => {
    const ht = new HashTable(4);
    ht.insert("isbn1", { title: "Old Title" });
    ht.insert("isbn1", { title: "New Title" }); // Update
    const result = ht.search("isbn1");
    strictEqual(result.title, "New Title", "Should update value");
})();

// Test: Delete
(() => {
    const ht = new HashTable(4);
    ht.insert("isbn1", { title: "To Delete" });
    ok(ht.has("isbn1"), "Key should exist before delete");
    
    const deleted = ht.delete("isbn1");
    strictEqual(deleted, true, "Delete should return true");
    strictEqual(ht.search("isbn1"), undefined, "Value should be gone");
    strictEqual(ht.length, 0, "Count should decrease");
})();

// Test: Search missing key
(() => {
    const ht = new HashTable(4);
    strictEqual(ht.search("missing"), undefined, "Missing key should return undefined");
})();

// Test: Resize on high load factor
(() => {
    const ht = new HashTable(4);
    const keys = ['a', 'b', 'c', 'd', 'e']; // Trigger resize at ~3 items
    keys.forEach(k => ht.insert(k, { x: k }));
    
    strictEqual(ht.size, 8, "Size should double after threshold");
    strictEqual(ht.length, 5, "All items should be preserved");
})();

// Test: Collision handling via separate chaining
(() => {
    const ht = new HashTable(4);
    ht.insert("a", { title: "A" });
    ht.insert("b", { title: "B" }); // May collide depending on hash
    
    strictEqual(ht.search("a").title, "A", "First book should be retrievable");
    strictEqual(ht.search("b").title, "B", "Second book should be retrievable");
})();

// Test: length and has
(() => {
    const ht = new HashTable(4);
    ht.insert("k1", { x: 1 });
    ht.insert("k2", { x: 2 });
    
    strictEqual(ht.length, 2, "Length should be 2");
    ok(ht.has("k1"), "Should contain k1");
    ok(ht.has("k2"), "Should contain k2");
    ok(!ht.has("k3"), "Should not contain k3");
})();

// Test: empty table
(() => {
    const ht = new HashTable(4);
    strictEqual(ht.length, 0, "Empty table length is 0");
    strictEqual(ht.count, 0, "Count should be 0");
    strictEqual(ht.loadFactor, 0, "Load factor should be 0");
    strictEqual(ht.search("any"), undefined, "Search should return undefined");
})();

console.log("\n✅ All tests passed!");